development/cpp/internal_rendering_architecture.rst

```diff
@@ -0,0 +1,694 @@
+    usage will increase in the shader. This happens even if all pixels evaluate
+    to ``true`` or ``false`` in a given frame.
+
+    If you use ``#if`` preprocessor branching, the number of required shader
+    versions will increase in the scene. In a worst-case scenario, adding a
+    single boolean ``#define`` can *double* the number of shader versions that
+    may need to be compiled in a given scene. In some cases, Vulkan
+    specialization constants can be used as a faster (but more limited)
+    alternative.
+
+    This means there is a high barrier to adding new built-in material features
+    in Godot, both in the core shaders and BaseMaterial3D. While BaseMaterial3D
+    can make use of dynamic code generation to only include the shader code if
+    the feature is enabled, it'll still require generating more shader versions
+    when these features are used in a project. This can make shader compilation
+    stutter more noticeable in complex 3D scenes.
+
+    See
+    `The Shader Permutation Problem <https://therealmjp.github.io/posts/shader-permutations-part1/>`__
+    and
+    `Branching on a GPU <https://medium.com/@jasonbooth_86226/branching-on-a-gpu-18bfc83694f2>`__
+    blog posts for more information.
+
+**Core GLSL material shaders:**
+
+- Clustered Forward: `servers/rendering/renderer_rd/shaders/scene_forward_clustered.glsl <https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/shaders/scene_forward_clustered.glsl>`__
+- Forward Mobile: `servers/rendering/renderer_rd/shaders/scene_forward_mobile.glsl <https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/shaders/scene_forward_mobile.glsl>`__
+- Compatibility: `drivers/gles3/shaders/scene.glsl <https://github.com/godotengine/godot/blob/master/drivers/gles3/shaders/scene.glsl>`__
+
+**Material shader generation:**
+
+- `scene/resources/material.cpp <https://github.com/godotengine/godot/blob/master/scene/resources/material.cpp>`__
+
+**Other GLSL shaders for Clustered Forward and Forward Mobile backends:**
+
+- `servers/rendering/renderer_rd/shaders/ <https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/shaders/>`__
+- `modules/lightmapper_rd/ <https://github.com/godotengine/godot/blob/master/modules/lightmapper_rd>`__
+
+**Other GLSL shaders for the Compatibility backend:**
+
+- `drivers/gles3/shaders/ <https://github.com/godotengine/godot/blob/master/drivers/gles3/shaders/>`__
+
+2D and 3D rendering separation
+------------------------------
+
+.. note::
+
+    Only available in the Clustered Forward and Forward Mobile backends, not the
+    Compatibility backend. Multiple Viewports can still be used to emulate this
+    when using the Compatibility backend, or to perform 2D resolution scaling.
+
+2D and 3D are rendered to separate buffers to allow for 3D resolution scaling
+without requiring the use of multiple Viewport nodes (which has greater
+performance and memory usage overhead).
```

Comment:
Done.

PS: I've wondered if we should say "SDR" instead of "LDR", as I see "SDR" much more often in popular parlance (such as on [monitor review sites](https://rtings.com/)). I rarely see LDR even mentioned in comparison. Do other engines refer to it as "LDR" or "SDR"?

Reply:
I've never heard the term "SSR" before, so I can't really say